packages/runtime/src/next/middleware/draft-middleware.ts

```
@@ -0,0 +1,89 @@
+}
+
+const SearchParams = {
+  DraftMode: 'x-makeswift-draft-mode',
+  PreviewMode: 'x-makeswift-preview-mode',
+}
+
+function getSecretFromRequest(request: NextRequest): string | null {
+  const draftModeSecret =
+    request.nextUrl.searchParams.get(SearchParams.DraftMode) ??
+    request.headers.get(HeaderNames.DraftMode)
+
+  const previewModeSecret =
+    request.nextUrl.searchParams.get(SearchParams.PreviewMode) ??
+    request.headers.get(HeaderNames.PreviewMode)
+
+  const isDraftModeRequest = draftModeSecret != null
+  const isPreviewModeRequest = previewModeSecret != null
+
+  return match({ isDraftModeRequest, isPreviewModeRequest })
+    .with({ isDraftModeRequest: true }, () => draftModeSecret)
+    .with({ isPreviewModeRequest: true }, () => previewModeSecret)
+    .otherwise(() => null)
+}
+
+function createProxyRequest(
+  request: NextRequest,
+  bypassResponse: BypassCookiesResponse,
+): NextRequest {
+  const proxyUrl = request.nextUrl.clone()
+  proxyUrl.searchParams.delete(SearchParams.DraftMode)
+  proxyUrl.searchParams.delete(SearchParams.PreviewMode)
+
+  // https://github.com/vercel/next.js/issues/52967#issuecomment-1644675602
+  // if we don't pass request twice, headers are stripped
+  const proxiedRequest = new NextRequest(request, request)
+
+  proxiedRequest.headers.delete(HeaderNames.DraftMode)
+  proxiedRequest.headers.delete(HeaderNames.PreviewMode)
+
+  bypassResponse.cookies.forEach(({ name, value }) => {
+    proxiedRequest.cookies.set(name, value)
+  })
+
+  return proxiedRequest
+}
+
+export function withDraftMiddleware(
+  { apiKey }: { apiKey: string },
+  middleware: NextMiddleware,
+): NextMiddleware {
+  return async (request, event) => {
+    const secret = getSecretFromRequest(request)
+
+    if (secret == null) {
+      return await middleware(request, event)
+    } else if (secret != null && secret !== apiKey) {
+      console.error('Incorrect Makeswift middleware secret, ignoring')
```

Do we want to ignore in this case as opposed to failing with a 4XX error?

Generally, I think making errors be handle-able or not swallowing them makes it easier to avoid confusing from end-users

---

For this and the error below, I'm fine with throwing errors. We've been burned by error throwing in the past, but those incidences are mostly related to rendering, not something as high level as a missing/incorrect secret.